extensions/amp-animation/0.1/web-animations.js

```diff
@@ -122,30 +136,80 @@ export class WebAnimationRunner {
     return this.playStateChangedObservable_.add(handler);
   }
 
+  /**
+   * @return {string}
+   */
+  createCodeBlob_() {
+
+    return `
+    registerAnimator('anim${++animIdCounter}', class {
+      animate(currentTime, effect) {
+        // TODO: Do some work with \`viewport-margins\` here.
+        if (currentTime == NaN) {
+          return;
+        }
+        effect.localTime = currentTime;
+      }
+    });
+    `;
+  }
+
   /**
    * Initializes the players but does not change the state.
    */
   init() {
     dev().assert(!this.players_);
-    this.players_ = this.requests_.map(request => {
-      // Apply vars.
-      if (request.vars) {
-        setStyles(request.target,
-            assertDoesNotContainDisplay(request.vars));
-      }
-      const player = request.target.animate(request.keyframes, request.timing);
-      player.pause();
-      return player;
-    });
-    this.runningCount_ = this.players_.length;
-    this.players_.forEach(player => {
-      player.onfinish = () => {
-        this.runningCount_--;
-        if (this.runningCount_ == 0) {
-          this.setPlayState_(WebAnimationPlayState.FINISHED);
+    if (this.useAnimationWorklet_) {
```

Comment:
could we move all of AnimationWorkled code to a separate file and treat is a "player" with a different type? Sort of like what https://github.com/ampproject/amphtml/pull/9194/files#diff-4f237a6703fccaa94b76ca8e9ae22d8e did (that code is gone from master since I later changed the approach) but the idea was AnimationRunner can work with different players and animationWorklet is a player

Reply:
Done, but kind of differently than you suggested. 

I made a base class called `AnimationRunner` and extended both `WebAnimationRunner` & `AnimationWorkletRunner` from it. I create one dependent on the `useAnimationWorklet_` flag which is now available on `Builder` instead. 

Does this look ok?